src/signalrclient/cancellation_token_source.h

```diff
@@ -94,7 +94,7 @@ namespace signalr
         void reset()
         {
             std::lock_guard<std::mutex> lock(m_lock);
-            assert(m_callbacks.empty());
+            //assert(m_callbacks.empty());
```

Comment:
?

Reply:
> Is it okay to reuse hub connection instance or should I recreate it after disconnection?

Yes.

> This assertion fails if I try to call start hub connection after it has disconnected.

That generally means something wasn't properly cleaned up.

Reply:
I will look into my app code and prepare a test case and then propose a fix.

Reply:
This happened then hub_connection disconnects on initial transport failure, for example then server is unavailable, and after that app calls start again, in my case this is a reconnect scenario, I have modified existing test to model this issue: 
```
TEST(start, propogates_exception_from_negotiate_and_starts_again)
{
    auto http_client = std::make_shared<test_http_client>([](const std::string& url, http_request, cancellation_token) -> http_response
        {
            throw custom_exception();
        });

    auto websocket_client = create_test_websocket_client();
    auto hub_connection = hub_connection_builder::create("http://fakeuri")
        .with_logging(std::make_shared<memory_log_writer>(), trace_level::none)
        .with_http_client_factory([http_client](const signalr_client_config& config)
            {
                http_client->set_scheduler(config.get_scheduler());
                return http_client;
            })
        .with_websocket_factory([websocket_client](const signalr_client_config&) { return websocket_client; })
        .build();

    auto mre = manual_reset_event<void>();
    hub_connection.start([&mre](std::exception_ptr exception)
        {
            mre.set(exception);
        });

    try
    {
        mre.get();
        ASSERT_TRUE(false);
    }
    catch (const custom_exception& e)
    {
        ASSERT_STREQ("custom exception", e.what());
    }

    hub_connection.start([&mre](std::exception_ptr exception)
        {
            
        });

    ASSERT_EQ(connection_state::disconnected, hub_connection.get_connection_state());
}
```
This happens because in the `connection_impl::start_negotiate`, around 202 line, `m_disconnect_cts` is not canceled like in the `connection_impl::shutdown`.

I think I should create separate issue on this matter because it is not directly connected with issue we solving here.